lamagraph-compiler/src/Lamagraph/Compiler/Eval.hs

```diff
@@ -0,0 +1,177 @@
+module Lamagraph.Compiler.Eval (evalCoreBindsDefEnv) where
+
+import Relude
+
+import Data.HashMap.Lazy qualified as HashMap
+
+import Control.Monad.Except
+import Lamagraph.Compiler.Core
+import Lamagraph.Compiler.Syntax
+import Lamagraph.Compiler.Typechecker.TcTypes
+
+data UnaryPrim = UPMinus | UPPrintInt deriving (Show)
+
+data BinaryPrim = BPPlus | BPMinus | BPTimes | BPLess deriving (Show)
+
+data Value
+  = VInt Int
+  | VChar Char
+  | VString Text
+  | VClosure Var CoreExpr EvalEnv
+  | VUnaryPrim UnaryPrim
+  | VBinaryPrim1 BinaryPrim
+  | VBinaryPrim2 BinaryPrim Value
+  | VTuple Value (NonEmpty Value)
+  | VAdt DataCon [Value]
+  deriving (Show)
+
+newtype EvalEnv = EvalEnv (HashMap Var Value) deriving (Show)
+
+data EvalError
```

Comment:
I'm not sure whether it's good to use `ExceptT` here. Maybe it'd better to use [Control.Exception](https://hackage.haskell.org/package/base-4.21.0.0/docs/Control-Exception.html)